src/sia/models/sia_query_params.py

```diff
@@ -0,0 +1,411 @@
+    "SIAQueryParams",
+    "Shape",
+    "DPType",
+    "Polarization",
+    "CalibLevel",
+]
+
+T = TypeVar("T", bound=Enum)
+
+
+class CalibLevel(int, Enum):
+    """Enumeration of possible calibration levels."""
+
+    LEVEL0 = 0
+    LEVEL1 = 1
+    LEVEL2 = 2
+    LEVEL3 = 3
+
+
+class Shape(CaseInsensitiveEnum):
+    """Enumeration of possible shapes for the POS parameter."""
+
+    CIRCLE = "CIRCLE"
+    RANGE = "RANGE"
+    POLYGON = "POLYGON"
+
+
+class DPType(CaseInsensitiveEnum):
+    """Enumeration of possible data product types."""
+
+    IMAGE = "IMAGE"
+    CUBE = "CUBE"
+
+
+class Polarization(CaseInsensitiveEnum):
+    """Enumeration of possible polarization states."""
+
+    I = "I"  # noqa: E741
+    Q = "Q"
+    U = "U"
+    V = "V"
+    RR = "RR"
+    LL = "LL"
+    RL = "RL"
+    LR = "LR"
+    XX = "XX"
+    YY = "YY"
+    XY = "XY"
+    YX = "YX"
+
+
+class BaseQueryParams(ABC):
+    """Base class for query parameters."""
+
+    @abstractmethod
+    def to_butler_parameters(self) -> Any:
+        """Convert the query parameters to the format expected by the
+        specific Butler type.
+        """
+
+
+@dataclass
+class SIAQueryParams(BaseQueryParams):
+    """A class to represent the parameters for an SIA query.
+
+    Attributes
+    ----------
+    pos : Optional[list[str]]
+        Positional region(s) to be searched.
+    q_format : Optional[list[str]]
+        Image format(s).
+    time : Optional[list[str]]
+        Time interval(s) to be searched.
+    band : Optional[list[str]]
+        Band interval(s) to be searched.
+    pol : Optional[list[Polarization]]
+        Polarization state(s) to be searched.
+    fov : Optional[list[str]]
+        Range(s) of field of view.
+    spatres : Optional[list[str]]
+        Range(s) of spatial resolution.
+    exptime : Optional[list[str]]
+        Range(s) of exposure times.
+    timeres : Optional[list[str]]
+        Range(s) of temporal resolution.
+    specrp : Optional[list[str]]
+        Range(s) of spectral resolving power.
+    q_id : Optional[list[str]]
+        Identifier of dataset(s). (Case insensitive)
+    dptype : Optional[list[DPType]]
+        Type of data (dataproduct_type).
+    calib : Optional[list[CalibLevel]]
+        Calibration level of the data.
+    target : Optional[list[str]]
+        Name of the target.
+    collection : Optional[list[str]]
+        Name of the data collection.
+    facility : Optional[list[str]]
+        Name of the facility.
+    instrument : Optional[list[str]]
+        Name of the instrument.
+    maxrec : Optional[int]
+        Maximum number of records in the response.
+    responseformat : Optional[str]
+        Format of the response.
+
+    Notes
+    -----
+    I have tried using Pydantic here, but as I understand there is currently
+    a limitation with how FastAPI handles Pydantic models for query parameters
+    with list attributes
+    (See: https://github.com/fastapi/fastapi/discussions/10556)
```

Comment:
I've created two classes as recommended to handle the different handlers. However both are still dataclasses, were you suggesting changing to Pydantic models? When I do try that I run into the issue where if I try to pass in list params (i.e. multiple of the same query param) the param ends up empty. I think this is https://github.com/fastapi/fastapi/discussions/10556 unless I've missunderstand or am misusing it?

Reply:
My suggestion is to just write a FastAPI dependency for GET and, in the body of that function, construct the model that's used for POST. See, for example, how the UWS library support in Safir recommends doing this, although it constructs a key/value pair model instead of a proper model like you'd want to construct.

https://safir.lsst.io/user-guide/uws/define-inputs.html#sync-get